### PR TITLE
[SPARK-47371] [SQL] XML: Ignore row tags found in CDATA

### PR DESCRIPTION
**Pull request by yhosny**
### What changes were proposed in this pull request?

The current parser does not recognize CDATA sections and thus will read row tags that are enclosed within a CDATA section. The expected behavior is for none of the following rows to be read, but they are all read. 

```
// BUG:  rowTag in CDATA section
val xmlString="""<?xml version="1.0" encoding="UTF-8" ?>
<test><![CDATA[
    <elem id="1" />
    <elem id="2" > </elem>
    <elem> <id>3</id> </elem>
]]>
</test>
```

The changes proposed would make the tokenizer ignore the row elements within the CDATA section. 

### Why are the changes needed?

For XML data to be read correctly.

### Does this PR introduce _any_ user-facing change?

rowTags within CDATA tags in XML files are now properly ignored when read by Spark.

### How was this patch tested?

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlParser.scala [hunk 1]
```diff
@@ -682,25 +684,25 @@ class XmlTokenizer(
         return false
       }
       val c = cOrEOF.toChar
-      if (c == commentEnd(i)) {
-        if (i >= commentEnd.length - 1) {
-          // Found comment close.
+      if (c == end(i)) {
+        i += 1
+        if (i >= end.length) {
```
Comment (sandip-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlParser.scala:689]: Please add a test with two scenarios:
- CDATA ends at the end of the file,
- CDATA never ends. 
The later will be invalid XML. Goal is to make sure the parser doesn't crash and still returns other valid records.

Add the same two tests for comments as well.

**Review by sandip-db — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala [hunk 2]
```diff
@@ -2584,6 +2584,21 @@ class XmlSuite
 
     val expectedResults = Seq.range(1, 18).map(Row(_))
     checkAnswer(results, expectedResults)
+
+    val results2 = spark.read.format("xml")
+      .option("rowTag", "ROW")
+      .option("multiLine", "true")
```
Comment (sandip-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala:2590]: nit: remove

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala [hunk 3]
```diff
@@ -2584,6 +2584,21 @@ class XmlSuite
...
+      .load(getTestResourcePath(resDir + "cdata-ending-eof.xml"))
+
+    val expectedResults2 = Seq.range(1, 18).map(Row(_))
+    checkAnswer(results2, expectedResults2)
+
+    val results3 = spark.read.format("xml")
+      .option("rowTag", "ROW")
+      .option("multiLine", "true")
```
Comment (sandip-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala:2598]: nit: remove

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala [hunk 4]
```diff
@@ -2584,6 +2584,21 @@ class XmlSuite
...
+      .load(getTestResourcePath(resDir + "cdata-no-close.xml"))
+
+    val expectedResults3 = Seq.range(1, 18).map(Row(_))
```
Comment (sandip-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala:2601]: Add checkAnswer(results3, expectedResults3)

**Review by sandip-db — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlParser.scala (hunk 1 above)
Comment (yhosny) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/xml/StaxXmlParser.scala:689]: Thanks for the review and test cases!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala (hunk 4 above)
Comment (yhosny) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/xml/XmlSuite.scala:2598]: Nice catch 👍

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**